core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java

```diff
@@ -87,4 +102,59 @@ public static TypeAnalyzer createTestingTypeAnalyzer(PlannerContext plannerConte
                         new TablePropertyManager(),
                         new AnalyzePropertyManager()));
     }
+
+    private static class QueryScopedCachedTypeAnalyzer
+    {
+        private final Cache<NodeRef<Expression>, Type> typesCache = buildNonEvictableCache(CacheBuilder.newBuilder());
+        private PlannerContext plannerContext;
+        private StatementAnalyzerFactory statementAnalyzerFactory;
+
+        private QueryScopedCachedTypeAnalyzer(PlannerContext plannerContext, StatementAnalyzerFactory statementAnalyzerFactory)
+        {
+            this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+            this.statementAnalyzerFactory = requireNonNull(statementAnalyzerFactory, "statementAnalyzerFactory is null");
+        }
+
+        private Map<NodeRef<Expression>, Type> getTypes(Session session, TypeProvider inputTypes, List<Expression> expressions)
+        {
+            List<NodeRef<Expression>> expressionsToResolve = collectExpressions(expressions);
```

Comment:
IMO it's overly complicated for what essentially is a call to `io.trino.sql.analyzer.ExpressionAnalyzer#analyze`.
E.g. each `getType` will go through this path (e.g. if we wanted to use it from `ExpressionInterpreter` when it's cached).